### ISSUE TITLE
Cyclic dependencies

### ISSUE THREAD

**Issue by tjdbryan** [nus-cs2030/2223-s2#331]
Hi all, I received this message on CodeCrunch.
![image](https://user-images.githubusercontent.com/122286169/223589885-ad506b30-cc26-4b8b-bd7a-bda8ecbf77cf.png)

Any idea how to resolve this issue?

Thanks!

**Comment by weitatftw** [comment]
there is a cyclic dependency between your Task and your TaskInterface, you would need to either have Task not rely on TaskInterface or the other way around to get rid of this error

**Comment by Nicknzk** [comment]
yup as @weitatftw stated, u need to ensure that ur (TaskInterface does not create a Task) or (Task not create a TaskInterface)

It seems like you are doing this for the level 5 view problem since the ImList<Task> cannot take RT so u want a ImList<TaskInterface> 

I recommend looking at overloading (theres some subthreads below regarding that)

**Comment by tjdbryan** [comment]
@weitatftw Thanks for pointing out!

**Comment by tjdbryan** [comment]
@Nicknzk you're absolutely right! I am trying to sort the ImList<TaskInterface> right before printing out using Comparator<Task> defined for earlier levels... Is there any way to circumvent this issue? Which method are you referring to for overloading? Thanks! 😊

For my design, Task implements TaskInterface.

**Comment by leonardwx** [comment]
I also faced the similar issues where I had multiple tasks replying on one another. What I did was to remove the redundant classes and merge some methods (overloading) where one class is able to do it

**Comment by sparklingwater13** [comment]
For me, I faced a similar problem as well. What I did was to check if my ImList takes in the TaskInterface ImList<TaskInterface> instead of ImList<Task>. Basically my RecurringTask, CancelledTask, CancelledRecurringTask, and Task all inherit from TaskInterface.

**Comment by tjdbryan** [comment]
@leonardwx Hmm, for my Planner, I'm using: `add(TaskInterface taskInstance)`, where Task, CancelledTask, RecurringTask, CancelledRecurringTask all inherit from TaskInterface. Thanks for the suggestion!

**Comment by tjdbryan** [comment]
@sparklingwater13 Hi, my design is similar to yours, Task, CancelledTask, RecurringTask, CancelledRecurringTask all inherit from TaskInterface. Can I check how you implemented the sorting of the ImList in ViewDay, right before iterating through the sorted list and printing out each task? Did you use the TaskComparator defined in the previous levels? Thanks!

**Comment by jarenl0ke** [comment]
@tjdbryan I've also faced a similar issue and in the end what I felt worked for me best was to not have the RecurringTask and Task be "on the same level" and inheriting from the same interface but rather try making the RecurringTask inherit from like Task (for example) and sorting this Task instead of sorting the TaskInterface itself. I have tried other methods (admittedly way too long on them) but this seems the most efficient way instead of adding too much extra code. 

Not sure how your other classes look like but this is roughly what worked for me and hope it helps..

**Comment by easypzlemonsqueezy** [comment]
your TaskInterface and Task classes depend on each other, you used Task within TaskInterface and TaskInterface within Task! if I'm not mistaken you should not use Task within TaskInterface, hope this helps!

**Comment by sparklingwater13** [comment]
> @sparklingwater13 Hi, my design is similar to yours, Task, CancelledTask, RecurringTask, CancelledRecurringTask all inherit from TaskInterface. Can I check how you implemented the sorting of the ImList in ViewDay, right before iterating through the sorted list and printing out each task? Did you use the TaskComparator defined in the previous levels? Thanks!

Hello, I defined another TaskComparator (lets call it Comp2) instead of using the one I defined in the previous level since the previous one compares Task objects. Comp2 compares TaskInterface objects instead. 

In my DayView, I added the Tasks I want into a new ImList<TaskInterface>, which are tasks on the same day and not cancelled. Remember to extract out the tasks from RecurringTasks too. Then I used Comp2 to sort the tasks based on their start times.

**Comment by littlemidget3** [comment]
thanks for asking this question! clarified a few doubts i had regarding my PA as well. My various types of task also inherit from a singular parent interface, thus it was important for me to check if my tasks were returning each other. I used the subthreads on overloading to counter my issues

**Comment by tjdbryan** [comment]
@jarenl0ke Thanks! The way that worked for me was to implement a new TaskInterfaceComparator that will help to sort RecurringTask & Task. If this design is closer to your lab design, you can consider using it! 😊

**Comment by tjdbryan** [comment]
@sparklingwater13 You're spot on! I defined a new Comparator and it worked. Thanks so much for your help! 😊

**Comment by tjdbryan** [comment]
@littlemidget3 Hi, what worked for me was to implement another Comparator for the overarching class/interface. Hope it works for you too. Thanks! 😊

**Comment by EvanCh08** [comment]
A tip for you to counter cyclic dependency is that make a spreadsheet and for each interface/class, you list down its dependency.

**Comment by jacoblimjy** [comment]
Remember to ensure that if class A relies on class B to be created, class B should not rely on class A to be created.

**Comment by tjdbryan** [comment]
Awesome! Thanks all for replying. You're all Top Gs!😉

**Comment by SimonSeeYongsheng** [comment]
I would recommend that you draw out your inheritance chart and check which methods are causing the cyclic dependency and take the appropriate steps to solve it

**Comment by huwenjie912** [comment]
i think perhaps there could some cyclic dependency between your task and taskinterface, i recommend looking overloading. Try listing out all the dependecies first then see if there are any loops

**Comment by stxe09** [comment]
What I did to overcome my cyclic dependency was create an additional interface, instead of extending from parent class, the child now implements the same interface as the original parent. But I think it might require you to make quite a bit of modification...

**Comment by e0968868** [comment]
I drew out all the classes to visualise how they are related to each other, i think this can help you to ensure that there arent any mutually dependent classes

**Comment by tjdbryan** [comment]
@reinatng10 not sure if that is allowed on CodeCrunch!

**Comment by tjdbryan** [comment]
@SimonSeeYongsheng @huwenjie912 @stxe09 @e0968868 thanks all. Today I learnt about inheritance chart 😊 BT2102 is useful too!
